### DISCUSSION TITLE
New Gatsby Image plugin does not expose originalName?

### DISCUSSION THREAD

**Discussion by LarsEjaas** [gatsbyjs/gatsby#30212] [Help] [answered]
Hi everybody

I am working on a migration from gatsby-image to gatsby-plugin-image, but I can't seem to wrap my head around this?

I have a lot of fluid images and use an image component where I can just write the filename of the image i need and not worry about the file location at all.

But I struggle a bit here - on the old plugin the "filename" was exposed inside the "originalName" field - but I can't seem to find anything similar within the new plugin?

This is my old image component:

```
import Img from "gatsby-image"

const Image = ({ fileName, alt, style, aspectR }) => {
  const { allImageSharp } = useStaticQuery(graphql`
    query {
      allImageSharp {
        nodes {
          fluid(maxWidth: 1408,quality: 90) {
            originalName
            ...GatsbyImageSharpFluid_withWebp
          }
        }
      }
    }
  `)

  const fluid = allImageSharp.nodes.find( node => node.fluid.originalName === fileName)
    .fluid 

  if (!fluid) return null;

  return (
    <>
      <Img fluid={fluid} alt={alt} style={style} />
    </>
  )
}

export default Image
```

Inside a component I would the just write my images like this to avoid any GraphQL queries inside the component:

`<Image  fileName="awesomePhoto.png" alt="this picture is great" />`

Using the GraphQL explorer I am looking inside "AllImageSharp > edges > node" And have "GatsbyImageData" but nothing similar to "OriginalName". 

Anyone found a solution for this?

**Comment by LarsEjaas** [comment, marked as answer]
I took a brake and gave it another go - I think I found a valid solution if anyone else need something similar.

```
const Image = ({ fileName, alt, loading }) => {
  const { allImageSharp } = useStaticQuery(graphql`
     {
      allImageSharp {
        nodes {
          gatsbyImageData(
            width: 550
            )
          parent {
            ... on File {
              name
            }
          }
        }
      }
    }
  `)

  const image = allImageSharp.nodes.find(node => node.parent.name === fileName).gatsbyImageData; 

  if (!image) return null;

  return (
    <>
      <GatsbyImage image={image} alt={alt} loading={loading} />
    </>
  )
}

export default Image
```

Please note: I use typeScript, so I might have missed something here when rewriting to JS? "loading" is "eager" or "lazy" depending on if the image is placed above/below the fold. FileName is specified without filename extension. E.g. "thisAwesomePhoto.png" should be written as "thisAwesomePhoto"

**Comment by LekoArts** [comment]
While this works you really shouldn't do it. Every time you use this component the page that contains it will load all images you have (as you query for all images and that data is inside the chunk). Please use the new `<StaticImage />` and programmatically pass images into `<GatsbyImage />`

**Reply by LarsEjaas** [reply]
Hi Lennart. Thanks for taking the time to respond :-) 

How do you then handle responsive images in Gatsby? I am probably missing something vital, but I can't seem to wrap my head around this with the new plugin?

**Reply by LekoArts** [reply]
What do you mean with "responsive images"?

**Reply by LarsEjaas** [reply]
Allright. I took a further read on the new image plugin. I totalled misunderstood how it works! 

"Static" images is a reference to the sourcing and has nothing to do with sizing. DOOH! 

I think "responsive images" is a widely used term - you know: "scale to fit the container". 

Oh well, Gatsby just got a LOT easier to use regarding images - I just need  "StaticImage" and "Constrained". I had to read the documentation a couple of times, I don't know, the wording in the description really wasn't intuitive to me!

Oh well, I hadn't figured this out without your help. Thanks!

**Reply by pascaloliv** [reply]
Hello I'm joining this discussion because I have the same workaround to grab the  `gatsbyImageData`.
I have to use `<GatsbyImage>` in Storybook and use some images mocks in my stories with the IGatsbyImageData structure. 

I'm building my UI with stateless components - generally with image props typed as GatsbyImageProps.
I can't use `<StaticImage>` out of the box because it accept only a string path (now we can't even pass [string as props](https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-plugin-image/#restrictions-on-using-staticimage)

I don't know if there is a better workaround for this.

**Reply by LarsEjaas** [reply]
Hi Olivier.

Well it works kind of alright, but it could lead to some really long build times for your page.  If you have a lot of images you probably have to be careful dividing them into folders and then create a separate instance at gatsby source filesystem for each folder like:

```
{
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `reference name`,
        path: `${__dirname}/src/articles/folder/`,
      },
    },
```

You can then limit your graphQL query to the images in the folder by filtering by "sourceInstanceName"

```
 allFile(filter: { sourceInstanceName: { eq: "reference name" } }) {
        edges {
          node {
```

But it is important to be aware that all images in the folder will be searched through on every build for every image on the page. 
To me it isn't that important really - I am on netlify and they have a nice plugin that will cache all images and only rebuild images if anything has changed.

**Reply by Defite** [reply]
> While this works you really shouldn't do it. Every time you use this component the page that contains it will load all images you have (as you query for all images and that data is inside the chunk). Please use the new `<StaticImage />` and programmatically pass images into `<GatsbyImage />`

Suppose I have `.mdx` document (not inside `pages` dir) and I want to put some component with image there. Like that

**My mdx:**

```
---
title: Some super cool text
---

<MyComponent image="/uploads/picForThisPost.jpg" alt="Super pic for super post" />
```

**My custom post tempalate `.tsx`:**

```
import React from 'react';
import { GatsbyImage, StaticImage } from 'gatsby-plugin-image';

const MyComponent: React.FunctionComponent = ({ image }) => {
  return (
   <>
     <GatsbyImage image={image} alt="" /> <-- won't work
     <StaticImage src={image} alt="" /> <-- won't work either
   </>
  )
}

export default MyComponent;
```

So what should I do to make this work? I want to pass image string to my component in mdx file and pass it through gatsby-plugin-image.

I can use static image from `/static/` folder but it's becoming to complicating to have multiple folders for images.

**Reply by LekoArts** [reply]
You could use `gatsby-remark-images` if it's only about displaying images and not `gatsby-plugin-image` specifically. About MDX: https://github.com/gatsbyjs/gatsby/discussions/27950#discussioncomment-364659

So you could open a feature request about it here: https://github.com/gatsbyjs/gatsby/discussions/categories/ideas-feature-requests

**Reply by Defite** [reply]
It's not only the display of image, but also about dir where to store images for post. If i put my image near my mdx file, I can't do something like:

```
<img src="./myimage.png" /> 
```

because it will search for file relative to root of my domain, like `localhost:8000/myimage.png` and will give me 404 error.

**Comment by agtz92** [comment]
Hey! I was looking at your solution and came across a property called getSrc.
[https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-plugin-image/#getsrc](url)

Just in case you were looking to simplify your code a little bit
Cheers!
